src/Microsoft.Health.Fhir.Shared.Api/Controllers/FhirController.cs

```diff
@@ -445,6 +447,46 @@ public async Task<IActionResult> ConditionalPatch(string typeParameter, [FromBod
             return ToSaveOutcomeResult(saveOutcome);
         }
 
+        /// <summary>
+        /// Patches the specified resource.
+        /// </summary>
+        /// <param name="typeParameter">The type.</param>
+        /// <param name="idParameter">The identifier.</param>
+        /// <param name="paramsResource">The JSON FHIR Parameters Resource.</param>
+        /// <param name="ifMatchHeader">Optional If-Match header.</param>
+        [HttpPatch]
+        [Route(KnownRoutes.ResourceTypeById)]
+        [AuditEventType(AuditEventSubType.Patch)]
+        [Consumes("application/fhir+json")]
+        public async Task<IActionResult> PatchFhir(string typeParameter, string idParameter, [FromBody] Parameters paramsResource, [ModelBinder(typeof(WeakETagBinder))] WeakETag ifMatchHeader)
+        {
+            UpsertResourceResponse response = await _mediator.PatchResourceAsync(new ResourceKey(typeParameter, idParameter), paramsResource, ifMatchHeader, HttpContext.RequestAborted);
+            return ToSaveOutcomeResult(response.Outcome);
+        }
+
+        /// <summary>
+        /// Conditionally patches a specified resource.
+        /// </summary>
+        /// <param name="typeParameter">Type of resource to patch.</param>
+        /// <param name="paramsResource">The JSON FHIR Parameters Resource.</param>
+        /// <param name="ifMatchHeader">Optional If-Match header.</param>
+        [HttpPatch]
+        [Route(KnownRoutes.ResourceType)]
+        [AuditEventType(AuditEventSubType.ConditionalPatch)]
+        [Consumes("application/fhir+json")]
+        public async Task<IActionResult> ConditionalPatchFhir(string typeParameter, [FromBody] Parameters paramsResource, [ModelBinder(typeof(WeakETagBinder))] WeakETag ifMatchHeader)
+        {
+            IReadOnlyList<Tuple<string, string>> conditionalParameters = GetQueriesForSearch();
+
+            UpsertResourceResponse response = await _mediator.Send<UpsertResourceResponse>(
+                new ConditionalFhirPatchResourceRequest(typeParameter, paramsResource, conditionalParameters, ifMatchHeader),
```

Comment:
>ConditionalFhirPatchResourceRequest [](http://example.com/codeflow?start=20&length=35)

This seems excessive for me. That whole class does same thing as ConditionalPatchRequest, with only difference is what is payload it carries.

I would do following:


Either make `PatchResourceRequest` generic by payload (either Parameters or JsonPatchDocument or just have two properites inside and two different constructors. 

Or make `PatchResourceRequest` an abstract class and derive two classes out of it FhirJson and Json

Rename JsonPatchService into PatchService and let it have either two public patch methods, or one public patch which would accept ConditionalPatchRequest and two private methods to handle patching. 

It feels like too much same code right now.

Reply:
I templated this and reduced complexity